server/src/main/java/org/elasticsearch/cluster/coordination/ElectionSchedulerFactory.java

```diff
@@ -0,0 +1,203 @@
+            ", maxTimeout=" + maxTimeout +
+            '}';
+    }
+
+    private class ElectionScheduler implements Releasable {
+        /**
+         * The current maximum timeout: the next election is scheduled randomly no later than this number of milliseconds in the future. On
+         * each election attempt this value is increased by `backoffTime`, up to the `maxTimeout`, to adapt to higher-than-expected latency.
+         */
+        private final AtomicLong currentMaxTimeoutMillis = new AtomicLong(Math.min(initialTimeout.millis(), maxTimeout.millis()));
+        private final AtomicBoolean isClosed = new AtomicBoolean();
+
+        /**
+         * Calculate the next maximum timeout by backing off the current value by `backoffTime` up to the `maxTimeout`.
+         */
+        private long backOffCurrentMaxTimeout(long currentMaxTimeoutMillis) {
+            return Math.min(maxTimeout.getMillis(), currentMaxTimeoutMillis + backoffTime.getMillis());
+        }
+
+        void scheduleNextElection(final TimeValue gracePeriod, final Runnable scheduledRunnable) {
+            if (isClosed.get()) {
+                logger.debug("{} not scheduling election", this);
+                return;
+            }
+
+            final long maxDelayMillis = currentMaxTimeoutMillis.getAndUpdate(this::backOffCurrentMaxTimeout);
+            final long delayMillis = randomPositiveLongLessThan(random::nextLong, maxDelayMillis + 1) + gracePeriod.millis();
+            final Runnable runnable = new AbstractRunnable() {
+                @Override
+                public void onFailure(Exception e) {
+                    logger.debug("unexpected exception in wakeup", e);
+                    assert false : e;
+                }
+
+                @Override
+                protected void doRun() {
+                    if (isClosed.get()) {
+                        logger.debug("{} not starting election", this);
+                        return;
+                    }
+                    logger.debug("{} starting election", this);
+                    scheduledRunnable.run();
+                }
+
+                @Override
+                public void onAfter() {
+                    scheduleNextElection(TimeValue.ZERO, scheduledRunnable);
+                }
+
+                @Override
+                public String toString() {
+                    return "scheduleNextElection{gracePeriod=" + gracePeriod
+                        + ", maxDelayMillis=" + maxDelayMillis
+                        + ", delayMillis=" + delayMillis
+                        + ", " + ElectionScheduler.this + "}";
+                }
+
+                @Override
+                public boolean isForceExecution() {
+                    // There are very few of these scheduled, and they back off, but it's important that they're not rejected as
+                    // this could prevent a cluster from ever forming.
+                    return true;
```

Comment:
still not sure this is needed. Have you checked this? If not, let's add a TODO here.

Reply:
`ThreadPool#schedule()` wraps this up in a `ThreadedRunnable`, but I think that's what we want. When the `ThreadedRunnable` finally runs, it runs this `AbstractRunnable` directly using the executor, so rejections seem to be handled right.

As far as I can see there's nothing special about whether the generic threadpool does rejections, but perhaps I'm missing something.

Reply:
The generic threadpool has an unbounded queue, so rejections should not happen unless the executor is shutdown. In that case, there's no need for this code to still execute.